pkg/apis/servicecatalog/validation/instance.go

```diff
@@ -335,35 +337,63 @@ func ValidateServiceInstanceReferencesUpdate(new *sc.ServiceInstance, old *sc.Se
-		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServiceClassName"), "exactly one of clusterServiceClassExternalName or clusterServiceClassName required"))
+	// Must specify exactly one source of the class: external id, external name, k8s name.
+	if (b2i(externalClassNameSet) + b2i(externalClassIDSet) + b2i(k8sClassSet)) != 1 {
+		classSetErrMsg := "exactly one of clusterServiceClassExternalName, clusterServiceClassExternalID, or clusterServiceClassName required"
+		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServiceClassExternalName"), classSetErrMsg))
+		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServiceClassExternalID"), classSetErrMsg))
+		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServiceClassName"), classSetErrMsg))
 	}
-	// Can't specify both External and k8s name but must specify one.
-	if externalPlanSet == k8sPlanSet {
-		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServicePlanExternalName"), "exactly one of clusterServicePlanExternalName or clusterServicePlanName required"))
-		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServicePlanName"), "exactly one of clusterServicePlanExternalName or clusterServicePlanName required"))
+	// Must specify exactly one source of the plan: external id, external name, k8s name.
+	if (b2i(externalPlanNameSet) + b2i(externalPlanIDSet) + b2i(k8sPlanSet)) != 1 {
+		planSetErrMsg := "exactly one of clusterServicePlanExternalName, clusterServicePlanExternalID, or clusterServicePlanName required"
+		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServicePlanExternalName"), planSetErrMsg))
+		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServicePlanExternalID"), planSetErrMsg))
+		allErrs = append(allErrs, field.Required(fldPath.Child("clusterServicePlanName"), planSetErrMsg))
 	}
 
-	if externalClassSet {
+	if externalClassNameSet {
 		for _, msg := range validateCommonServiceClassName(p.ClusterServiceClassExternalName, false /* prefix */) {
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("clusterServiceClassExternalName"), p.ClusterServiceClassExternalName, msg))
 		}
 
 		// If ClusterServiceClassExternalName given, must use ClusterServicePlanExternalName
-		if !externalPlanSet {
+		if !externalPlanNameSet {
 			allErrs = append(allErrs, field.Required(fldPath.Child("clusterServicePlanExternalName"), "must specify clusterServicePlanExternalName with clusterServiceClassExternalName"))
 		}
 
 		for _, msg := range validateCommonServicePlanName(p.ClusterServicePlanExternalName, false /* prefix */) {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("clusterServicePlanExternalName"), p.ClusterServicePlanName, msg))
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("clusterServicePlanExternalName"), p.ClusterServicePlanExternalName, msg))
+		}
+	}
+	if externalClassIDSet {
```

Comment:
shouldn't this be an `else if`? if name isn't set, seems like we'd need to check ID instead, not in addition to